crates/ir/src/container.rs

```diff
@@ -0,0 +1,107 @@
+use std::mem;
+
+use crate::{
+    op_code::{IrInstruction, IrOpCode},
+    register::{chill::RegisterChill, RegisterId},
+};
+
+pub struct IrContainer {
+    _data: Vec<u8>,
+}
+
+impl IrContainer {
+    /// # Safety
+    /// This function can receive any data and it is up to the caller to ensure that the data is valid IR in valid
+    /// version.
+    pub unsafe fn from_vec(data: Vec<u8>) -> Self {
+        IrContainer { _data: data }
+    }
+
+    pub fn iter(&self) -> IrContainerIterator {
+        IrContainerIterator {
+            container: self,
+            index: 0,
+        }
+    }
+}
+
+impl<'container> IntoIterator for &'container IrContainer {
+    type Item = IrInstruction<'container>;
+    type IntoIter = IrContainerIterator<'container>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+pub struct IrContainerIterator<'container> {
+    container: &'container IrContainer,
+    index: usize,
+}
+
+impl<'container> IrContainerIterator<'container> {
+    /// # Safety
+    /// Caller must ensure that the data is valid IR in valid version.
+    unsafe fn read_instruction(&mut self) -> IrInstruction<'container> {
+        match self.read_op_code() {
+            // ! Local Memory
```

Comment:
Again, I cannot see anything that's `unsafe` about this function. Remember that `unsafe` does not mean "this function may do things which may result in unsafe code being called *down the line*, but itself is implemented in 100% safe code." `unsafe` is pretty strictly just for functions that *may themselves actually perform Undefined Behavior.*